pkg/fga/client.go

```diff
@@ -0,0 +1,942 @@
+		req.Header.Set("Warrant-Token", opts.WarrantToken)
+	}
+
+	res, err := c.HTTPClient.Do(req)
+	if err != nil {
+		return CheckResponse{}, err
+	}
+	defer res.Body.Close()
+
+	if err = workos_errors.TryGetHTTPError(res); err != nil {
+		return CheckResponse{}, err
+	}
+
+	var checkResponse CheckResponse
+	dec := json.NewDecoder(res.Body)
+	err = dec.Decode(&checkResponse)
+	if err != nil {
+		return CheckResponse{}, err
+	}
+
+	return checkResponse, nil
+}
+
+func (c *Client) CheckBatch(ctx context.Context, opts CheckBatchOpts) ([]CheckResponse, error) {
+	c.once.Do(c.init)
+
+	checkOpts := CheckOpts{
+		Op:           CheckOpBatch,
+		Checks:       opts.Checks,
+		Debug:        opts.Debug,
+		WarrantToken: opts.WarrantToken,
+	}
+	data, err := c.JSONEncode(checkOpts)
+	if err != nil {
+		return []CheckResponse{}, err
+	}
+
+	endpoint := fmt.Sprintf("%s/fga/v1/check", c.Endpoint)
+	req, err := http.NewRequest(http.MethodPost, endpoint, bytes.NewBuffer(data))
+	if err != nil {
+		return []CheckResponse{}, err
+	}
+
+	req = req.WithContext(ctx)
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "Bearer "+c.APIKey)
+	req.Header.Set("User-Agent", "workos-go/"+workos.Version)
+	if opts.WarrantToken != "" {
+		req.Header.Set("Warrant-Token", opts.WarrantToken)
+	}
+
+	res, err := c.HTTPClient.Do(req)
+	if err != nil {
+		return []CheckResponse{}, err
+	}
+	defer res.Body.Close()
+
+	if err = workos_errors.TryGetHTTPError(res); err != nil {
+		return []CheckResponse{}, err
+	}
+
+	var checkResponses []CheckResponse
+	dec := json.NewDecoder(res.Body)
+	err = dec.Decode(&checkResponses)
+	if err != nil {
+		return []CheckResponse{}, err
+	}
+
+	return checkResponses, nil
+}
+
+// Query executes a query for a set of resources.
+func (c *Client) Query(ctx context.Context, opts QueryOpts) (QueryResponse, error) {
+	c.once.Do(c.init)
+
+	endpoint := fmt.Sprintf("%s/fga/v1/query", c.Endpoint)
+	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
+	if err != nil {
+		return QueryResponse{}, err
+	}
+
+	req = req.WithContext(ctx)
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "Bearer "+c.APIKey)
+	req.Header.Set("User-Agent", "workos-go/"+workos.Version)
+	if opts.WarrantToken != "" {
+		req.Header.Set("Warrant-Token", opts.WarrantToken)
+	}
+
+	if opts.Limit == 0 {
+		opts.Limit = ResponseLimit
+	}
+
+	if opts.Order == "" {
+		opts.Order = Desc
+	}
+
+	type QueryUrlOpts struct {
+		Query        string `url:"q"`
+		Context      string `url:"context,omitempty"`
+		Limit        int    `url:"limit,omitempty"`
+		Order        Order  `url:"order,omitempty"`
+		Before       string `url:"before,omitempty"`
+		After        string `url:"after,omitempty"`
+		WarrantToken string `url:"-"`
+	}
+
+	var jsonCtx []byte
+	if opts.Context != nil {
+		jsonCtx, err = json.Marshal(opts.Context)
+		if err != nil {
+			return QueryResponse{}, err
+		}
+	}
+	queryUrlOpts := QueryUrlOpts{
```

Comment:
Rather than defining this inline struct to accommodate context being a `string` vs a `Context` (so that the conversion to url params is done correctly), we should define an `EncodeValues` method on the `QueryOpts` struct (so it implements the [`query.Encoder`](https://pkg.go.dev/github.com/google/go-querystring/query#Encoder) interface) which converts the `Context` to a `string` upon conversion to url values.